zebra-network/src/peer_set/candidate_set.rs

```diff
@@ -329,24 +332,49 @@ where
 /// If the data in an address is invalid, this function can:
 /// - modify the address data, or
 /// - delete the address.
-//
-// TODO: re-enable this lint when last_seen_limit is used
-#[allow(unused_variables)]
+///
+/// Currently, this method will offset the reported `last_seen` time to prevent clock skews
+/// from causing the peers to be placed too far back or in the front of the reconnection queue
+/// incorrectly.
 fn validate_addrs(
     addrs: impl IntoIterator<Item = MetaAddr>,
     last_seen_limit: DateTime<Utc>,
-) -> impl IntoIterator<Item = MetaAddr> {
+) -> impl Iterator<Item = MetaAddr> {
     // Note: The address book handles duplicate addresses internally,
     // so we don't need to de-duplicate addresses here.
 
     // TODO:
     // We should eventually implement these checks in this function:
-    // - Zebra should stop believing far-future last_seen times from peers (#1871)
     // - Zebra should ignore peers that are older than 3 weeks (part of #1865)
     //   - Zebra should count back 3 weeks from the newest peer timestamp sent
     //     by the other peer, to compensate for clock skew
     // - Zebra should limit the number of addresses it uses from a single Addrs
     //   response (#1869)
 
-    addrs
+    let mut addrs: Vec<_> = addrs.into_iter().collect();
+
+    if !addrs.is_empty() {
+        limit_last_seen_times(&mut addrs, last_seen_limit);
+    }
+
+    addrs.into_iter()
+}
+
+/// Ensure all reported `last_seen` times are less than or equal to `last_seen_limit`.
+///
+/// This function assumes there is at least one address in the `addrs` list.
+fn limit_last_seen_times(addrs: &mut Vec<MetaAddr>, last_seen_limit: DateTime<Utc>) {
+    let most_recent_reported_seen_time = addrs
+        .iter()
+        .map(|addr| addr.get_last_seen())
+        .max()
+        .expect("at least one peer in the address list");
+
+    let offset = last_seen_limit - most_recent_reported_seen_time;
+
+    for addr in addrs {
+        if addr.get_last_seen() > last_seen_limit {
+            addr.offset_last_seen_by(offset);
+        }
+    }
```

Comment:
If an honest peer has future clock skew, we'll want to reduce all of its times by the same amount.

For malicious peers, this is also a slight security improvement. If they have any addresses that are before the current time, they also get moved further back in the queue.

(Each response from a single peer is validated separately.)

```suggestion
    if most_recent_reported_seen_time > last_seen_limit {
        let offset = last_seen_limit - most_recent_reported_seen_time;
        for addr in addrs {
            addr.offset_last_seen_by(offset);
        }
    }
```